File: src/Turtle.java
/**
 * Simulates a turtle object. Must be extended to use.
 *
 * @author Dafu Ai
 */
abstract class Turtle implements Tickable {
    /**
     * The patch this turtle is currently at.
     */
    protected Patch patch;

    /**
     * The world this turtle is in.
     */
    protected World world;

    /**
     * Construct a new Turtle instance.
     * @param world world this turtle is in.
     */
    Turtle(World world) {
        this.world = world;

        if (canMove()) {
            move();
        }
    }

    /**
     * Move to a random, unoccupied patch.
     */
    private void move() {
        Patch newPatch = world.getPatchMap().getRandomUnoccupiedPatch();
        moveToPatch(newPatch);
    }

    /**
     * Move to a specified patch.
     * @param newPatch the new patch to be moved to
     */
    void moveToPatch(Patch newPatch) {
        if (patch != null) {
            patch.removeTurtle(this);
        }
        newPatch.addTurtle(this);
        this.patch = newPatch;
    }

    /**
     * Get the current patch.
     * @return the current patch
     */
    Patch getPatch() {
        return patch;
    }

    /**
     * Update the current state
     */
    public void update() {
        move();
    }

    /**
     * Determines whether this turtle can move.
     * @return true if it can; false otherwise
     */
    abstract boolean canMove();
}

File: rebellion-python/models/Turtle.py
from models import World, Patch

# Author: Dafu Ai


class Turtle:
    """
    Simulates a turtle object, in which the behaviours are shared by both Cop and Turtle.
    """
    world: World    # The world this turtle is in.
    patch: Patch    # The patch this turtle is currently at.

    def __init__(self, world: World) -> None:
        """Place itself to a patch."""
        self.world = world
        self.patch = None
        self.move()

    def can_move(self) -> bool:
        """Determines whether this turtle can move."""
        return True

    def move_to_patch(self, new_patch: Patch) -> None:
        """Move to a specified patch."""

        if self.patch is not None:
            self.patch.remove_turtle(self)

        new_patch.add_turtle(self)
        self.patch = new_patch

    def move(self) -> None:
        """Move to a random, unoccupied patch."""
        if not self.can_move():
            return

        new_patch = self.world.patch_map.get_random_unoccupied_patch()
        self.move_to_patch(new_patch)

    def update(self) -> None:
        """Update the current state by moving to another place"""
        self.move()
